NDArrayMap2 and NDArrayReindex

johnc1231 opened the pull request:
This PR:

- Introduces `NDArrayReindex`, which allows the use of the `transpose` on hail `NDArrayExpressions`. 
- Introduces `NDArrayMap2`, which allows operations like elementwise +, -, /, *, etc. This has full broadcasting support to match numpys. 
- Fixes `NDArrayShape`, which was not properly implemented. The only reason `NDArrayShape` seemed to work was because calling `NDArrayShape` directly on a `MakeNDArray` or on a `NDArrayMap1` was always simplified away to just return the tuple. 
- Removed the Simplify rule that transformed `NDArrayShape(MakeNDArray(..., shape))` to `shape`, since that was resulting in a type change, as `NDArrayShape` always returns a required tuple and `shape` was not a required tuple.
- Changes `ndarrays_eq` in the tests to just go through `eval` instead of a write then read, since write and read don't exist yet. 

Things I'm not happy with:
- The rule for emitting `NDArrayReindex` is written twice, since depending on the situation one or the other is computationally cheaper. Seems like there should be a version of `NDArrayEmitter` that allows us to use an already existing array as the data. 

Might add a few more tests, but code is here for review.

danking commented on hail/src/main/scala/is/hail/expr/types/physical/PNDArray.scala:
```diff
@@ -36,7 +36,7 @@ final case class PNDArray(elementType: PType, nDims: Int, override val required:
       .loadField(shape, idx))
 
     if (nDims == 0) {
-      0L
```
was this uncovered by removing an optimizer rule?  If not, we should add a test for this case.

johnc1231 replied:
There are tests for this. This is for 0 dimensional ndarrays, like 

`hl._ndarray(5)` or `np.array(5)`. They have 0 dimensions and their shape is the empty tuple.

danking replied:
What you're telling me is that:
```
Array.range(0, nDims).foldLeft(const(1L)) {...}
```
Will do exactly the right thing then.

johnc1231 replied:
Yup! Good catch. It was a separate case when I didn't understand the need to support 0 dimensional NDArrays, and I didn't realize the condition could be removed now that we acknowledge 0 dimensional NDArrays as having 1 element.

danking commented:
Aside: we should work towards making smaller PRs

tpoterba commented on hail/src/main/scala/is/hail/expr/ir/Optimize.scala:
```diff
@@ -26,7 +26,7 @@ object Optimize {
 
     if (ir.typ != ir0.typ)
-      fatal(s"optimization changed type!\n  before: ${ ir0.typ }\n  after:  ${ ir.typ }" +
+      fatal(s"optimization changed type!\n  before: ${ ir0.typ.parsableString() }\n  after:  ${ ir.typ.parsableString() }" +
```
yes, good change! Definitely wrong to print the py type here

johnc1231 commented on hail/src/main/scala/is/hail/expr/ir/Emit.scala:
```diff
@@ -2014,7 +2202,7 @@ abstract class NDArrayEmitter(
    val setup: Code[_]) {
 
-  def outputElement(idxVars: Seq[ClassFieldRef[Long]]): Code[_]
+  def outputElement(idxVars: Seq[Settable[Long]]): Code[_]
```
I can go back to `ClassFieldRef`s. I got a comment at some point about making a lot of class fields so I was trying to avoid making tons of class fields by using locals. I don't understand tradeoffs here in general (though what you're saying about it getting split makes sense).

johnc1231 commented on hail/python/test/hail/expr/test_expr.py:
```diff
@@ -3212,13 +3193,12 @@ def test_ndarray_map_jvm(self):
-                                           [-5, -6, -7]]))
-        assert(np.array_equal(hl.eval(c), [[True, True, True],
-                                           [True, True, True]]))
+        self.ndarrays_eq(
```
Great, I'm going to split out a `test_ndarray.py` file. NDArrays are going to have tons of tests, deserves its own file.

johnc1231 commented on hail/src/main/scala/is/hail/expr/ir/Emit.scala:
```diff
@@ -1991,21 +2040,160 @@ private class Emit(
+        val rElemRef = coerce[Any](mb.newField(rName)(rVti))
+
+        val bodyEnv = env.bind(name=lName, v=(lVti, false, lElemRef.load()))
+                         .bind(name=rName, v=(rVti, false, rElemRef.load()))
```
Maybe I was just being silly, but I was having a problem using the varargs version of bind with exactly two arguments, as it was complaining it couldn't decide which overloading of bind I meant.

danking replied:
ah, possible.

johnc1231 commented on hail/src/main/scala/is/hail/expr/ir/Emit.scala:
```diff
@@ -1414,13 +1413,63 @@ private class Emit(
+          reindexShapeAndStrides,
+          ndAddress := outputPType.construct(childFlags, childOffset, shapeSrvb.end(), stridesSrvb.end(), childDataAddress, mb)
+        )
+        EmitTriplet(setup, false, ndAddress)
```
So this is two things. To your first point, yeah, I should handle missingness here. To your second point about `setup` and `ndAddress`, I think I need that. In my experience, if I don't bind it to a name and I return a block of code, that block of code is run every time someone uses the `NDArray`. Like we rerun the entire code block over and over again everytime someone calls `.v` on the `EmitTriplet`.

danking replied:
This seems like a bug in the rest of the emitter, but I suppose that's what we're stuck with.

johnc1231 commented on hail/src/main/scala/is/hail/expr/ir/Emit.scala:
```diff
@@ -1414,13 +1413,63 @@ private class Emit(
+
+        val outputPType = x.pType.asInstanceOf[PNDArray]
+        val outputShapePType = outputPType.representation.fieldType("shape")
+        val outputStridesPType = outputPType.representation.fieldType("strides")
```
I love this, that's going to help a lot.

johnc1231 commented on hail/src/main/scala/is/hail/expr/ir/Emit.scala:
```diff
@@ -1414,13 +1413,63 @@ private class Emit(
-        val shape = t.loadField(region, ndt.value[Long], "shape")
+        val repr = ndIR.pType.asInstanceOf[PNDArray].representation
+
+        EmitTriplet(ndt.setup, false, repr.loadField(region, ndt.value[Long], "shape"))
```
How so? Isn't present for when there is no setup?

johnc1231 commented on hail/src/main/scala/is/hail/expr/ir/Emit.scala:
```diff
@@ -1414,13 +1413,63 @@ private class Emit(
+        EmitTriplet(setup, false, ndAddress)
+      case _: NDArrayMap | _: NDArrayMap2 =>
+        val emitter = emitDeforestedNDArray(resultRegion, ir, env)
+        emitter.emit(ir.pType.asInstanceOf[PNDArray])
```
I'm going to change the name of `emitDeforestedNDArray` to `deforestNDArray`, so it's not weird that that thing returns an `NDArrayEmitter`. Then I'm going to make the helper function called `emitDeforestedNDArray`

johnc1231 commented on hail/src/main/scala/is/hail/expr/ir/InferType.scala:
```diff
@@ -115,13 +115,14 @@ object InferType {
       case NDArrayShape(nd) =>
-        TTuple(nd.typ.required, List.tabulate(nd.typ.asInstanceOf[TNDArray].nDims)(_ => TInt64()):_*)
+        val ndType = nd.typ.asInstanceOf[TNDArray]
+        ndType.representation.fieldType("shape").asInstanceOf[TTuple].setRequired(ndType.required)
```
To do this, I'd have to replicate the StaticallyKnownField stuff on TNDArray as well, which I think I won't bother with given that we are moving towards dropping requiredness out of types anyway.

johnc1231 commented:
Alright, I think I've addressed just about everything. I didn't handle missingness in `NDArrayReindex`, as that's something I'm going to do in a future pass of general missingness handling

danking commented on hail/src/main/scala/is/hail/expr/ir/Emit.scala:
```diff
@@ -1417,15 +1419,63 @@ private class Emit(
+          reindexShapeAndStrides,
+          ndAddress := x.pType.construct(childFlags, childOffset, shapeSrvb.end(), stridesSrvb.end(), childDataAddress, mb)
+        )
+        EmitTriplet(setup, false, ndAddress)
```
OK, this method needs to ultimately look more like this:

```scala
val child = emit(childIR)
val childPType = child.pType.asInstanceOf[PNDArray]
val childOff = mb.newField[Long]
val childShape = new CodePTuple(
  childPType.shape.pType, region, childPType.shape.load(region, childOff)
val childShape = new CodePTuple(
  childPType.strides.pType, region, childPType.strides.load(region, childOff)
val setup = Code(
    child.setup,
    childM := child.m,
    childOff := childM.mux(defaultValue[Long], child.value[Long]))
val value = x.pType.construct(
  childPType.flags.load(region, childOff),
  childPType.offset.load(region, childOff),
  { srvb =>
    Code(indexExpr.map { index =>
      if (index < childPType.ndims) childShape(index) else 1L
    }.map(srvb.addLong _): _*)
  },
  { srvb =>
    Code(indexExpr.map { index =>
      if (index < childPType.ndims) childStrides(index) else 0L
    }.map(srvb.addLong _): _*)
  },
  _.addLong(childPType.data.load(region, childOff)))

EmitTriplet(setup, childM.load(), value)
```

Benefits:

- no unnecessary data copies
- the srvb lives in `PNDArray`, closer to the definition of all the types that
  make up a `PNDArray`, no need to duplicate that code elsewhere
- overall shorter, which aids readability

danking commented on hail/src/main/scala/is/hail/expr/ir/Emit.scala:
```diff
@@ -1996,21 +2046,139 @@ private class Emit(
+              lElemRef := leftChildEmitter.outputElement(lIdxVars2),
+              rElemRef := rightChildEmitter.outputElement(rIdxVars2),
+              bodyt.setup,
+              bodyt.m.orEmpty(Code._fatal("NDArray map body cannot be missing")),
```
We can't lay down the missingness bit and do nothing about it!

`EmitArrayTriplet` should probably not take `Option[Code[Boolean]]` but some `Missingness` type that has named alternatives. For now, remember that `None` means definitely not-missing (see `MakeArray`). If the missingness is `None`, everything is fine, we should not throw an error. If the missing ness is `Some(m)` we need to emit `m.mux(Code._fatal(...), Code._empty)`.

johnc1231 replied:
I don't use `EmitArrayTriplet` at all, a little confused by this comment.

danking replied:
Ah, disregard. I was misreading this code, now I understand what orEmpty does.

danking commented on hail/src/main/scala/is/hail/expr/ir/Emit.scala:
```diff
@@ -1996,21 +2046,139 @@ private class Emit(
+        val childPType = child.pType.asInstanceOf[PNDArray]
+
+        val outputPType = x.pType
+        val outputShapePType = outputPType.representation.fieldType("shape").asInstanceOf[PTuple]
```
use the new `.shape.typ`

danking commented on hail/src/main/scala/is/hail/expr/ir/Emit.scala:
```diff
@@ -1996,21 +2046,139 @@ private class Emit(
+
+        val setup = Code(childEmitter.setup, shapeSrvb.start(), reindexShape)
+
+        new NDArrayEmitter(mb, indexExpr.length, shapeSrvb.end(), outputShapePType, outputPType.elementType, setup) {
```
oof, sorry again, we can't be doing this.

We're not "deforesting" if every IR node could potentially allocate a shape tuple!

NDArrayEmitter probably should take an `SRVB => Code[Unit]` which will add the shape to the ndarray we're building.

danking replied:
`makeDefaultStrides` needs to be changed to use an SRVB as well, it shouldn't be manipulating the region directly like that.

danking replied:
OK, so I think each NDArrayEmitter should just define for itself what its shape *and* strides are. Trying to do it at the end is gonna be a mess because we want to look at the shape of an NDArray that we're currently building using an SRVB.

danking replied:
Can we make them `Array[Code[Long]]`? I think that's actually the right thing.

johnc1231 replied:
You want the shape and the strides to be passed along as `Array[Code[Long]]`, and then I only realize one shape tuple at the end? That should be fine.

johnc1231 replied:
Just to be clear, the elements of this array are Code's representing the longs that will make up the shape?

danking commented on hail/src/main/scala/is/hail/expr/ir/Emit.scala:
```diff
@@ -1996,21 +2046,139 @@ private class Emit(
+    val setup = Code(
+      unifiedShapeAddress := unifiedPType.allocate(region),
+      Code(setupList:_*))
+    (setup, unifiedShapeAddress, unifiedPType)
```
There's no reason to have a setup if you don't have missingness.

danking commented on hail/src/main/scala/is/hail/expr/ir/Emit.scala:
```diff
@@ -1996,21 +2046,139 @@ private class Emit(
+    }
+
+    val setup = Code(
+      unifiedShapeAddress := unifiedPType.allocate(region),
```
We should basically never call `allocate`. Always use an `SRVB`, it's safer.

danking commented on hail/src/main/scala/is/hail/expr/ir/Emit.scala:
```diff
@@ -1996,21 +2046,139 @@ private class Emit(
+    val unifiedShapeAddress = mb.newField[Long]
+
+    val setupList = Array.tabulate(unifiedPType.size) { idx =>
+      val left = getShapeAtIdx(leftShape, idx)
```
So, this is going to allocate a shape tuple for each dimension you look up. Which makes this calculation (which should be constant time in the number of dimensions and number of IR nodes) `O(number_of_dimensions^number_of_ir_nodes)`

danking replied:
We need to make the changes I mentioned earlier about `NDArrayEmitter`'s having `Array[Code[Long]]` I think that's the right way to fix all this. No allocations.

danking commented on hail/src/main/scala/is/hail/expr/ir/Emit.scala:
```diff
@@ -1996,21 +2046,139 @@ private class Emit(
+      Code(
+         notSameAndNotBroadcastable.mux(
+          Code._fatal("Incompatible NDArray shapes"),
+          Region.storeLong(unifiedPType.loadField(unifiedShapeAddress, idx),
```
If you find yourself reaching for `storeXXX`, you probably need an SRVB.

danking commented on hail/src/main/scala/is/hail/expr/ir/Emit.scala:
```diff
@@ -1996,21 +2046,139 @@ private class Emit(
+    val setupList = Array.tabulate(unifiedPType.size) { idx =>
+      val left = getShapeAtIdx(leftShape, idx)
+      val right = getShapeAtIdx(rightShape, idx)
+      val notSameAndNotBroadcastable = !((left ceq right) || (left ceq const(1L)) || (right ceq const(1L)))
```
does `1L` not work on its own? I'd think `ceq`'s type signature was sufficient to trigger the coercion.

johnc1231 commented on hail/src/main/scala/is/hail/asm4s/Code.scala:
```diff
@@ -933,3 +936,14 @@ class CodeNullable[T >: Null : TypeInfo](val lhs: Code[T]) {
     ifNull[U](Code._null[U], cnonnullcase)
 }
+
+class CodePTuple(
```
Didn't even know those were different things. Will do

johnc1231 replied:
I just put it in PTuple.scala

johnc1231 commented on hail/src/main/scala/is/hail/expr/ir/Emit.scala:
```diff
@@ -1379,23 +1383,21 @@ private class Emit(
-        val repr = xP.representation
+        val xP = x.pType
         val dataContainer = dataIR.pType.asInstanceOf[PArray]
         val shapePType = shapeIR.pType.asInstanceOf[PTuple]
```
I didn't even know you could use coerce that way. I've only used `coerce[Long]` with a Scala numerical type.

johnc1231 commented on hail/src/main/scala/is/hail/expr/ir/Emit.scala:
```diff
@@ -1417,15 +1419,63 @@ private class Emit(
         EmitTriplet(setup, false, ndAddress)
-      case x@NDArrayShape(ndIR) =>
+      case NDArrayShape(ndIR) =>
         val ndt = emit(ndIR)
```
I had seen t being used for "triplet"

johnc1231 commented on hail/src/main/scala/is/hail/expr/ir/Emit.scala:
```diff
@@ -1417,15 +1419,63 @@ private class Emit(
           )
```
Addressed by #7140

johnc1231 commented on hail/src/main/scala/is/hail/expr/ir/Emit.scala:
```diff
@@ -1417,15 +1419,63 @@ private class Emit(
+        val childt = emit(child)
+        val childPointer = mb.newField[Long]
+
+        val childPType = child.pType.asInstanceOf[PNDArray]
```
Going to save this comment for the future, don't think it'll make a big difference either way at this time after the rewrites.

johnc1231 commented:
Think I've addressed all your comments. I addressed missingness in reindex because you explicitly mentioned it, but the ndarray code overall is going to need a missingness overhaul where I add tests for missing values. I'll do that in a subsequent PR. I wasn't sure what your comment about `EmitArrayTriplet` was about, that's not something I use.

danking commented on hail/src/main/scala/is/hail/expr/types/physical/PNDArray.scala:
```diff
@@ -33,74 +54,66 @@ final case class PNDArray(elementType: PType, nDims: Int, override val required:
-      Array.range(0, nDims).foldLeft(const(1L)) { (prod, idx) => prod * getShapeAtIdx(idx) }
-    }
+  def numElements(shape: Array[Code[Long]], mb: MethodBuilder): Code[Long] = {
+      Array.range(0, nDims).foldLeft(const(1L)) { (prod, idx) => prod * shape(idx) }
```
now that we're using shape arrays this is just `shape.foldLeft(const(1L), _ * _)`

johnc1231 commented on hail/python/hail/expr/expressions/typed_expressions.py:
```diff
@@ -3279,7 +3279,7 @@ def transpose(self, axes=None):
         else:
             if len(axes) != self.ndim:
-                raise ValueError(f'Must specify a complete permutation of the dimensions.'
+                raise ValueError(f'Must specify a complete permutation of the dimensions. '
```
Did this on purpose. Otherwise it prints like: `Must specify a complete permutation of the dimensions.Expected ....`

johnc1231 commented on hail/src/main/scala/is/hail/expr/ir/Emit.scala:
```diff
@@ -2116,63 +2164,173 @@ private class Emit(
+        val ndAddress = mb.newField[Long]
+        val setup = Code(
+          ndt.setup,
+          ndAddress := ndt.value[Long]
```
Will address in subsequent PR.

danking replied:
I created and assigned https://github.com/hail-is/hail/issues/7221 to you

danking commented:
This is a big step! Good work John.
